src/chain_sync/network_context.rs

```diff
@@ -139,10 +140,33 @@ where
         &self,
         peer_id: Option<PeerId>,
         tsk: &TipsetKey,
-        count: u64,
+        count: NonZeroU64,
     ) -> Result<Vec<Arc<Tipset>>, String> {
-        self.handle_chain_exchange_request(peer_id, tsk, count, HEADERS, |_| true)
-            .await
+        self.handle_chain_exchange_request(
+            peer_id,
+            tsk,
+            count,
+            HEADERS,
+            |tipsets: &Vec<Arc<Tipset>>| {
+                if let Some(start) = tipsets.first() {
+                    if start.key() != tsk {
+                        tracing::warn!(epoch=%start.epoch(), expected=%tsk, actual=%start.key(), "start tipset key mismatch");
+                        return false;
+                    }
+                    for (ts, pts) in tipsets.iter().zip(tipsets.iter().skip(1)) {
+                        if ts.parents() != pts.key() {
+                            tracing::warn!(epoch=%ts.epoch(), expected_parent=%pts.key(), actual_parent=%ts.parents(), "invalid chain");
+                            return false;
+                        }
+                    }
+                    true
+                } else {
+                    tracing::warn!(%count, "invalid empty chain_exchange_headers response");
+                    false
+                }
+            },
```

Comment:
Would it make sense to extract this to a named method a write a unit test to check and document the behaviour? It also might prevent future regressions.

Reply:
fixed